src/printer.js

```diff
@@ -3154,6 +3174,16 @@ function printJSXElement(path, options, print) {
     }
   });
 
+  if (children[0] === innerJsxWhitespace) {
+    children[0] = leadingJsxWhitespace;
```

Comment:
No reason it needs to be done this way, it was just the quickest way for me to test whether the idea of having different leading/trailing whitespace would work.

Very happy to try some different approaches to this. I wonder if the simplest would be to flatten the `groups` array. That way we wouldn't need to duplicate the logic of converting inner whitespace to leading/trailing whitespace.